velox/functions/prestosql/StringFunctions.h

```diff
@@ -494,4 +494,64 @@ struct LevenshteinDistanceFunction {
   }
 };
 
+template <typename T>
+struct NormalizeFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // Map for holding normalization form options
+  const static inline std::unordered_map<std::string, utf8proc_int16_t>
+      normalizationOptions{
+          {"NFC", (UTF8PROC_STABLE | UTF8PROC_COMPOSE)},
+          {"NFD", (UTF8PROC_STABLE | UTF8PROC_DECOMPOSE)},
+          {"NFKC", (UTF8PROC_STABLE | UTF8PROC_COMPOSE | UTF8PROC_COMPAT)},
+          {"NFKD", (UTF8PROC_STABLE | UTF8PROC_DECOMPOSE | UTF8PROC_COMPAT)}};
+
+  FOLLY_ALWAYS_INLINE void initialize(
+      const std::vector<TypePtr>& /*inputTypes*/,
+      const core::QueryConfig& /*config*/,
+      const arg_type<Varchar>* /*string*/,
+      const arg_type<Varchar>* form) {
+    VELOX_USER_CHECK_NOT_NULL(form);
+    VELOX_USER_CHECK_NE(
+        normalizationOptions.count(*form),
+        0,
+        "Normalization form must be one of [NFD, NFC, NFKD, NFKC]");
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& string) {
+    doCall(result, string, "NFC");
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& string,
+      const arg_type<Varchar>& form) {
+    doCall(result, string, form);
+  }
+
+  // Note: This function newly allocates output using malloc so it should be
+  // free'd at the end.
+  FOLLY_ALWAYS_INLINE void doCall(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& string,
+      const arg_type<Varchar>& form) {
+    utf8proc_uint8_t* output = nullptr;
+    auto outputLength = utf8proc_map(
+        (utf8proc_uint8_t*)string.data(),
+        string.size(),
+        &output,
+        normalizationOptions.at(form));
+    if (outputLength < 0) {
+      result = string;
+    } else {
+      result.resize(outputLength);
+      std::memcpy(
+          result.data(), reinterpret_cast<const char*>(output), outputLength);
```

Comment:
could you add a check if result.data() is null and not call memcpy in that case (this happens when outputLength is 0)